src/Rule/RequiredHandler.php

```diff
@@ -31,15 +29,24 @@ public function validate(mixed $value, object $rule, ValidationContext $context)
         }
 
         $result = new Result();
+        if ($context->isAttributeMissing()) {
+            $translatedMessage = $this->translator->translate($rule->getNotPassedMessage());
+            $result->addError($translatedMessage);
+
+            return $result;
+        }
 
-        if ($this->isEmpty(is_string($value) ? trim($value) : $value)) {
-            $formattedMessage = $this->translator->translate(
-                $rule->getMessage(),
-                ['attribute' => $context->getAttribute(), 'value' => $value]
-            );
-            $result->addError($formattedMessage);
+        if (is_string($value)) {
+            $value = trim($value);
```

Comment:
Maybe make `trim` optional? 
Suggest make this in another PR.

Reply:
`trim` was already there before these changes. Can be seen as new because of the refactoring. Yes, suggest to improve it in a separate PR. I actually had it written in my notes, created the issue - #328.